Add dummy backend (#3820)
woodpecker-ci/woodpecker: docs/docs/92-development/09-testing.md

# Testing

## Backend

### Unit Tests

[We use default golang unit tests.](https://go.dev/doc/tutorial/add-a-test)
With [`"github.com/stretchr/testify/assert"`](https://pkg.go.dev/github.com/stretchr/testify@v1.9.0/assert) to simplify the test code.

### Integration Tests

#### Pipeline Engine

> The pipeline engine has a special backend called **`dummy`** witch does not exec but emulate how a typical backend should behave.

Author: ```suggestion
The pipeline engine has a special backend called **`dummy`** witch does not exec but emulate how a typical backend should behave.
To have it available you need to build the agent or cli with the `integration` tag.
```

# Testing

## Backend

### Unit Tests

> [We use default golang unit tests.](https://go.dev/doc/tutorial/add-a-test)
> With [`"github.com/stretchr/testify/assert"`](https://pkg.go.dev/github.com/stretchr/testify@v1.9.0/assert) to simplify testing.

Reviewer: ```suggestion
[We use default golang unit tests](https://go.dev/doc/tutorial/add-a-test)
with [`"github.com/stretchr/testify/assert"`](https://pkg.go.dev/github.com/stretchr/testify@v1.9.0/assert) to simplify testing.
```

### Integration Tests

> #### Pipeline Engine

Reviewer: ```suggestion
### Dummy backend
```

> The pipeline engine has a special backend called **`dummy`** which does not execute any commands, but emulates how a typical backend should behave.

Reviewer: ```suggestion
There is a special backend called **`dummy`** which does not execute any commands, but emulates how a typical backend should behave.
```

> To enable it you need to build the agent or cli with the `integration` build tag.

Reviewer: ```suggestion
To enable it you need to build the agent or cli with the `test` build tag.
```

[echo:L6:0s] ------------------
[echo:L7:0s] 9:18PM DBG pipeline/pipeline.go:176 > prepare CLI=exec step=echo
9:18PM DBG pipeline/pipeline.go:203 > executing CLI=exec step=echo
9:18PM TRC pipeline/backend/dummy/dummy.go:81 > start step echo taskUUID=01J10P578JQE6E25VV1EQF0745
9:18PM TRC pipeline/backend/dummy/dummy.go:167 > tail logs of step echo taskUUID=01J10P578JQE6E25VV1EQF0745
[echo:L0:0s] StepName: echo
[echo:L1:0s] StepType: commands
[echo:L2:0s] StepUUID: 01J10P578JQE6E25VV1DFSXX1Y
[echo:L3:0s] StepCommands:
[echo:L4:0s] ------------------
[echo:L5:0s] echo ja
[echo:L6:0s] ------------------
[echo:L7:0s] 9:18PM TRC pipeline/backend/dummy/dummy.go:108 > wait for step echo taskUUID=01J10P578JQE6E25VV1EQF0745
9:18PM TRC pipeline/backend/dummy/dummy.go:187 > stop step echo taskUUID=01J10P578JQE6E25VV1EQF0745
9:18PM DBG pipeline/pipeline.go:209 > complete CLI=exec step=echo
9:18PM TRC pipeline/backend/dummy/dummy.go:208 > delete workflow environment taskUUID=01J10P578JQE6E25VV1EQF0745
```

You can control the step behavior via its name:

- If you name it `step_start_fail` the engine will simulate a step start who fail (e.g. happens when the container image can not be pulled).

> - If you name it `step_exec_error` the engine will simulate a command who executes with status code **1**.

Reviewer: ```suggestion
- If you name the step `step_exec_error` the engine will simulate a command which executes with status code **1**.
```

> There are also environment variables to alter things:

Reviewer: ```suggestion
There are also environment variables to alter step behaviour:
```

- `SLEEP` witch will simulate a given time duration as command execution time.

> - `EXPECT_TYPE` witch let the backend error if set and the step is not the expected step-type.

Reviewer: ```suggestion
- `EXPECT_TYPE` allows to check if a step is a `step` or `service`
```

> - `EXPECT_TYPE` allows to check if a step is a `step` or `service`

Reviewer: ```suggestion
- `EXPECT_TYPE` allows to check if a step is a `clone`, `service`, `plugin` or `commands`
```

[echo:L7:0s] 9:18PM DBG pipeline/pipeline.go:176 > prepare CLI=exec step=echo
9:18PM DBG pipeline/pipeline.go:203 > executing CLI=exec step=echo
9:18PM TRC pipeline/backend/dummy/dummy.go:81 > start step echo taskUUID=01J10P578JQE6E25VV1EQF0745
9:18PM TRC pipeline/backend/dummy/dummy.go:167 > tail logs of step echo taskUUID=01J10P578JQE6E25VV1EQF0745
[echo:L0:0s] StepName: echo
[echo:L1:0s] StepType: commands
[echo:L2:0s] StepUUID: 01J10P578JQE6E25VV1DFSXX1Y
[echo:L3:0s] StepCommands:
[echo:L4:0s] ------------------
[echo:L5:0s] echo ja
[echo:L6:0s] ------------------
[echo:L7:0s] 9:18PM TRC pipeline/backend/dummy/dummy.go:108 > wait for step echo taskUUID=01J10P578JQE6E25VV1EQF0745
9:18PM TRC pipeline/backend/dummy/dummy.go:187 > stop step echo taskUUID=01J10P578JQE6E25VV1EQF0745
9:18PM DBG pipeline/pipeline.go:209 > complete CLI=exec step=echo
9:18PM TRC pipeline/backend/dummy/dummy.go:208 > delete workflow environment taskUUID=01J10P578JQE6E25VV1EQF0745
```

There are also environment variables to alter step behaviour:

- `SLEEP: 10` will let the step wait 10 seconds
- `EXPECT_TYPE` allows to check if a step is a `clone`, `service`, `plugin` or `commands`

> - `STEP_START_FAIL: true` if set will simulate a step to fail before actually started (e.g. happens when the container image can not be pulled)

Reviewer: ```suggestion
- `STEP_START_FAIL: true` if set will simulate a step to fail before actually being started (e.g. happens when the container image can not be pulled)
```

- `STEP_TAIL_FAIL: true` if set will error when we simulate to read from stdout for logs
- `STEP_EXIT_CODE: 2` if set will be used as exit code, default is 0

> - `STEP_OOM_KILLED: true` simulate a step who got killed because of memory constrains

Reviewer: ```suggestion
- `STEP_OOM_KILLED: true` simulates a step being killed by memory constrains
```

> Also for the whole workflow if you set it's UUID ot `WorkflowSetupShouldFail`, you can fail a workflow setup fail.

Reviewer: ```suggestion
You can let the setup of a whole workflow fail by setting it's UUID to `WorkflowSetupShouldFail`.
```
